### ISSUE TITLE
running out of buffers with recvmsg-multishot since 6.1.7

### ISSUE THREAD

**Issue by lano1106** [axboe/liburing#778]
Here are the settings:

8192 provided buffers

[2023-01-20 11:14:28] INFO WSCONT/initGroupBuffer Provide 8192 16724 bytes buffers to gid 1 from address 0x0x7f710bd57000 (137003008)

to serve 48 sockets recvmsg multishots.

I configure my ring with twice as much cqe (16K) to avoid cq overflows if all buffers get used.
I do use the buffer rings to return back buffers to io_uring.

Very shortly after the connections are initiated, data is exchanged back and forth, I see the HTTP handshake but in less than a minute, the program start to experience a non stop flow of recvmsg errors. ENOBUF (105).

If I rollback the kernel to 6.1.6, I have stable connections with no errors at all for hours.

Prior to 6.1.7, I may have seen 1 or 2 ENOBUF errors once or twice with this number of buffers (8192) after running the application non-stop for weeks. Now, I have a spurious flow of them. Hundreds of them in few minutes...

I suspect something broke with 6.1.7

**Comment by axboe** [comment]
Hmm interesting - looking at the log, these are the changes between 6.1.6 and 6.1.7:

```
Jens Axboe (5):
      io_uring/poll: add hash if ready poll request can't complete inline
      io_uring/fdinfo: include locked hash table in fdinfo output
      io_uring/poll: attempt request issue after racy poll wakeup
      io_uring/io-wq: free worker if task_work creation is canceled
      io_uring/io-wq: only free worker if it was allocated for creation

Pavel Begunkov (1):
      io_uring: lock overflowing for IOPOLL
```

Are you compiling your own kernels? If so, I'd probably just bisect around io_uring/ between those two commits. The two poll related changes would be primary candidates. Really don't see how that can be right now, but that doesn't mean there isn't an issue.

I'm assuming you don't have a test case for this that I can run...?

**Comment by lano1106** [comment]
it makes sense that 2 of the poll commits are primary candidates. Sure, I'll try bisect to pinpoint what is causing my issue.

No, I do not have a test case at this point and I have barely started to look at the problem...

it is not impossible too that a fix is now breaking my app. The fixed bug maybe had the effect of throttling down the incoming flow.

A feature of the flow is that despite providing very large buffers, the sending server do send very small segments most of the time (100-200) bytes on average.

Initially, my number of buffers was set to 512 but with multishot, it was depleted too fast. I had to double the size until I reach 8192.

I am going isolate the issue. With that result in hand maybe things will get clearer...

**Comment by axboe** [comment]
You could try this...

```
diff --git a/io_uring/poll.c b/io_uring/poll.c
index f2f9f174fc62..ab5ae475840f 100644
--- a/io_uring/poll.c
+++ b/io_uring/poll.c
@@ -281,8 +281,12 @@ static int io_poll_check_events(struct io_kiocb *req, bool *locked)
 			 * to the waitqueue, so if we get nothing back, we
 			 * should be safe and attempt a reissue.
 			 */
-			if (unlikely(!req->cqe.res))
+			if (unlikely(!req->cqe.res)) {
+				/* Multishot armed need not reissue */
+				if (!(req->apoll_events & EPOLLONESHOT))
+					continue;
 				return IOU_POLL_REISSUE;
+			}
 		}
 		if (req->apoll_events & EPOLLONESHOT)
 			return IOU_POLL_DONE;
```

**Comment by axboe** [comment]
I suspect that:

```
commit c06015ebc4367be38904b88582e13cc079672075
Author: Jens Axboe <axboe@kernel.dk>
Date:   Sat Jan 14 08:46:14 2023 -0700

    io_uring/poll: attempt request issue after racy poll wakeup
```

seems like the most likely culprit.

**Comment by lano1106** [comment]
thx... I am giving this potential fix a shot. I'll you know ASAP how it went

**Comment by axboe** [comment]
Wrote a test case (https://git.kernel.dk/cgit/liburing/commit/?id=15de013e2456dae5fce9c7bcb6711fd563d656a9) and verified it works with 6.1.6, and break miserably with 6.1.7. With the above mentioned fix, it works again as expected.

We might have another issue here with leaks of ring provided buffers, which is just apparent with the above bug. But that's for a separate investigation, I'll do that next.

**Comment by lano1106** [comment]
I confirm too that your patch did fix the behavior that I was observing with 6.1.7

I don't know if I'll run out of provided buffers. I'll let you know if I see something in that regard.

Something that I wish I had is access to is the buffer ring tail and head indexes for easier debug on the client side.

I think that I might have created some debug code having a bitmap to track down the bids returned by io_uring to figure out if I was mishandling the buffers...

this might be handy to reactivate that code to detect a potential leak...

**Comment by axboe** [comment]
Great, thanks for reporting and testing. I'll send out a patch for this.

I'm looking into any potential leaks now. Multishot should not leak with the fix, it definitely did before the fix on 6.1.7 (or current -git). Separate investigation... I'll close this one up.
